src/Main.elm

```diff
@@ -628,21 +653,108 @@ raisedDiv =
+viewModal model (DifficultyMenu difficulty fields) =
+    let
+        menuContent =
+            Html.map MenuMsg <|
+                div []
+                    [ radiobutton "Beginner" (difficulty == Beginner) Beginner
+                    , radiobutton "Intermediate" (difficulty == Intermediate) Intermediate
+                    , radiobutton "Expert" (difficulty == Expert) Expert
+                    , radiobutton "Custom" (isCustom difficulty) (Custom fields)
+                    , viewCustomFields (isCustom difficulty) fields
+                    ]
+    in
     div maskStyle
         [ modalContent []
-            [ windowsChrome [ style "padding" "0 18px 18px" ]
+            [ windowsChrome [ style "padding" "0 18px 90px" ]
                 [ formGroup "Difficulty"
-                    [ radiobutton "Beginner" Beginner model.game
-                    , radiobutton "Intermediate" Intermediate model.game
-                    , radiobutton "Expert" Expert model.game
+                    [ menuContent
+                    , button [ onClick (CloseMenu (Just difficulty)) ] [ text "OK" ]
+                    , button [ onClick (CloseMenu Nothing) ] [ text "Cancel" ]
                     ]
                 ]
             ]
         ]
 
 
+radiobutton : String -> Bool -> Difficulty -> Html MenuMsg
+radiobutton settingLabel isSelected difficulty =
+    label [ style "display" "flex", style "align-items" "center" ]
+        [ input
+            [ type_ "radio"
+            , name "value"
+            , onClick (SetDifficulty difficulty)
+            , checked isSelected
+            , style "margin" "4px 8px"
+            ]
+            []
+        , text settingLabel
+        ]
+
+
+viewCustomFields : Bool -> GridProperties -> Html MenuMsg
+viewCustomFields enabled ({ width, height, bombs } as fields) =
+    let
+        toInput num onInput_ =
+            input
+                [ type_ "number"
+                , value (String.fromInt num)
+                , onInput
+                    (String.toInt
+                        >> Maybe.withDefault 0
+                        >> onInput_
+                    )
+                , style "margin" "4px 8px"
+                , disabled (not enabled)
+                ]
+                []
+    in
+    div []
+        [ toInput width
+            (\num ->
+                SetDifficulty (Custom { fields | width = clamp 1 20 num })
+            )
+        , toInput height
+            (\num ->
+                SetDifficulty (Custom { fields | height = clamp 1 20 num })
+            )
+        , toInput bombs
+            (\num ->
+                SetDifficulty
+                    (Custom
+                        { fields
+                            | bombs =
+                                if num > width * height - 1 then
+                                    abs (width * height - 1)
+
+                                else
+                                    abs num
+                        }
```

Comment:
i'm doing the `clamp` stuff in the view itself. it's probably "better" to do this in `updateMenu`, since it would make it more impossible to end up with unwanted values. however, i thought it was more simple to implement it this way, and since we have the small `MenuMsg` type, i'm pretty confident we won't mess up and accidentally set unwanted values in our `Difficulty`. 🤷‍♂